presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -287,6 +292,45 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
             return new ExchangeNode(node.getId(), node.getType(), node.getScope(), partitioningScheme, sources, inputs);
         }
 
+        private Symbol canonicalizeExchangeNodeOutput(ExchangeNode node, int symbolIndex, Multimap<Symbol, Symbol> inputToOutputs)
+        {
+            Symbol canonicalOutput = canonicalize(node.getOutputSymbols().get(symbolIndex));
+            Optional<Symbol> outputSymbolForInputSymbols = getExchangeNodeOutputSymbolForInputSymbols(node, symbolIndex, inputToOutputs);
+
+            if (outputSymbolForInputSymbols.isPresent()
+                    && !canonicalOutput.equals(outputSymbolForInputSymbols.get())) {
+                map(canonicalOutput, outputSymbolForInputSymbols.get());
+                return outputSymbolForInputSymbols.get();
+            }
+            else {
+                node.getInputs().stream()
+                        .forEach(input -> inputToOutputs.put(input.get(symbolIndex), canonicalOutput));
```

Comment:
I  think `inputToOutputs` should contain canonical versions of input and output. That case you will also optimize output when inputs are different but their canonical versions are the same.